### PR TITLE
kubeadm: Replace MigrateOldConfigFromFile

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

`MigrateOldConfigFromFile` is a function, whose purpose is to migrate one config into another. It is working OK for now, but it has some issues:

- It is incredibly inefficient. It can reload and re-parse a single config file for up to 3 times.

- Because of the reloads, it has to take a file containing the configuration (not a byte slice as most of the rest config functions). However, it returns the migrated config in a byte slice (rather asymmetric from the input method).

- Due to the above points it's difficult to implement a proper interface for deprecated kubeadm config versions.

To fix the issues of `MigrateOldConfigFromFile`, the following is done:

- Re-implement the function by removing the calls to file loading package public APIs and replacing them with newly extracted package private APIs that do the job with pre-provided input data in the form of `map[GroupVersionKind][]byte`.

- Take a byte slice of the input configuration as an argument. This makes the function input symmetric to its output. Also, it's now renamed to `MigrateOldConfig` to represent the change from config file path as an input to byte slice.

- As a bonus (actually forgotten from a previous change) `BytesToInternalConfig` is renamed to the more descriptive `BytesToInitConfiguration`.

**Which issue(s) this PR fixes**:

Refs kubernetes/kubeadm#1296

**Special notes for your reviewer**:

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/assign @neolit123
/assign @fabriziopandini 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/util/config/joinconfiguration.go
Comment (neolit123) [inline cmd/kubeadm/app/util/config/joinconfiguration.go]: typo `betweek`

**Review by neolit123 — COMMENTED** [review]
thanks @rosti 
found one typo only.

the changes LGTM
/approve
/hold
for more comments by others if needed.

**Review by neolit123 — APPROVED** [review]
/lgtm
/retest

**Review by yagonobre — COMMENTED** [review]
Thanks @rosti 
/lgtm

**Review by neolit123 — APPROVED** [review]
/hold cancel
LGTM unholding to unblock https://github.com/kubernetes/kubernetes/pull/74025

**Pull request merged**
